test/testfile_helper.c

```diff
@@ -0,0 +1,301 @@
+            break;
+        (*lineno)++;
+        if ((p = strchr(buff, '\n')) == NULL) {
+            fprintf(stderr, "Line %d too long.\n", *lineno);
+            break;
+        }
+        *p = '\0';
+        frag = p - buff;
+
+        /* Append the physical line. */
+        copy = OPENSSL_realloc(line, linelen + frag + 1);
+        if (copy == NULL) {
+            fprintf(stderr, "Out of memory at line %d\n", *lineno);
+            break;
+        }
+        line = copy;
+        strcpy(&line[linelen], buff);
+        linelen += strlen(buff);
+
+        if (p == buff || p[-1] != '\\')
+            return line;
+        line[--linelen] = '\0';
+    }
+
+    /* Failed; clean up and return. */
+    free(line);
+    return NULL;
+}
+
+/* Decode %xx URL-decoding in-place; return 0 on failure. */
+static int urldecode(char *p)
+{
+    char *out = p;
+
+    for (; *p; p++) {
+        if (*p != '%')
+            *out++ = *p;
+        else if (isxdigit(p[1]) && isxdigit(p[2])) {
+            /* Don't check, can't fail because of ixdigit() call. */
+            *out++ = (OPENSSL_hexchar2int(p[1]) << 4)
+                   | OPENSSL_hexchar2int(p[2]);
+            p += 2;
+        }
+        else
+            return 0;
+    }
+    *out = '\0';
+    return 1;
+}
+
+STANZA *stanza_parse_fp(FILE *fp)
+{
+    STANZA *sp = NULL, *save = NULL;
+    PAIR *pp;
+    char *line, *equals, *key, *value;
+    int encoded, comment, lineno = 0;
+    size_t len;
+
+    /* Skip any prolog. */
+    for ( ; ; ) {
+        line = read_line(fp, &lineno);
+        if (line == NULL) {
+            fprintf(stderr, "Empty file.\n");
+            return NULL;
+        }
+        if (line[0] == '#' || line[0] == '\0') {
+            OPENSSL_free(line);
+            continue;
+        }
+        break;
+    }
+
+    save = sp = (STANZA *)OPENSSL_zalloc(sizeof(*sp));
+    sp->start = lineno;
+    for ( ; ; ) {
+        /* Name? */
+        if (line[0] == '[') {
+            OPENSSL_free(sp->name);
+            sp->name = OPENSSL_strdup(line + 1);
+            if ((equals = strrchr(sp->name, ']')) != NULL)
+                *equals = '\0';
+        } else {
+            /* Add a key/value pair line. */
+            encoded = 0;
+            if ((equals = strchr(line, '~')) != NULL)
+                encoded++;
+            else if ((equals = strchr(line, '=')) == NULL) {
+                fprintf(stderr, "Line %d missing equals.\n", sp->start);
+                return save;
+            }
+            *equals++ = '\0';
+
+            key = trim_spaces(line);
+            value = skip_spaces(equals);
+            if (*value == '"') {
+                value++;
+                len = strlen(value);
+                if (len < 2 || value[--len] != '"') {
+                    fprintf(stderr, "Line %d missing close quote.\n", sp->start);
+                    return save;
+                }
+                value[len] = '\0';
+            }
```

Comment:
Remove this `if` clause entirely and let value fetching do the unquoting *if* it wants a string or a bin.  That is, the day you decide to implement something like `stanza_get_bin` and possibly `stanza_get_string`, that would be the place to do the quotes (and possibly unescape) parsing.  But here, just take the values literally as they are, just with spaces trimmed at start and end.  I believe that will make a port of `evp_test.c` a bit easier...